File: backend/app/services/platform_addons.py
"""Estimate-level Databricks Platform add-on pricing."""

from __future__ import annotations

import json
import math
from datetime import date
from functools import lru_cache
from pathlib import Path
from typing import Any


CATALOG_PATH = (
    Path(__file__).resolve().parents[2]
    / "static"
    / "pricing"
    / "platform-addons.json"
)


@lru_cache(maxsize=1)
def load_platform_addon_catalog() -> dict[str, Any]:
    """Load the curated percentage-based Platform add-on catalog."""
    with CATALOG_PATH.open(encoding="utf-8") as catalog_file:
        return json.load(catalog_file)


def _pricing_date(value: date | str | None) -> date:
    if value is None:
        return date.today()
    if isinstance(value, date):
        return value
    return date.fromisoformat(value)


def get_selected_platform_addon(
    discount_config: dict[str, Any] | None,
) -> str | None:
    """Read the estimate-level add-on selection from its JSON envelope."""
    if not discount_config or not isinstance(discount_config, dict):
        return None
    selections = discount_config.get("platform_addons", [])
    if selections in (None, []):
        return None
    if not isinstance(selections, list):
        raise ValueError("discount_config.platform_addons must be a list")
    if len(selections) > 1:
        raise ValueError(
            "Select only one Platform add-on; Mission Critical already includes "
            "Enhanced Security and Compliance"
        )
    selection = selections[0]
    if not isinstance(selection, str):
        raise ValueError("Platform add-on identifiers must be strings")
    return selection.upper()


def get_platform_addon_discount(
    discount_config: dict[str, Any] | None,
) -> float:
    """Read the negotiated discount applied after the add-on uplift."""
    if not discount_config or not isinstance(discount_config, dict):
        return 0.0
    global_discounts = discount_config.get("global", {})
    if not isinstance(global_discounts, dict):
        return 0.0
    value = float(global_discounts.get("platform_addon_discount", 0) or 0)
    if not math.isfinite(value) or value < 0 or value > 100:
        raise ValueError(
            "global.platform_addon_discount must be between 0 and 100"
        )
    return value


def get_platform_addon_rate(
    addon_type: str,
    cloud: str,
    tier: str,
    pricing_date: date | str | None = None,
) -> dict[str, Any]:
    """Resolve availability and the active published uplift."""
    catalog = load_platform_addon_catalog()
    addon_key = (addon_type or "").upper()
    cloud_key = (cloud or "").upper()
    tier_key = (tier or "").upper()
    addon = catalog["addons"].get(addon_key)
    if addon is None:
        raise ValueError(f"Unknown Platform add-on: {addon_type}")

    cloud_config = addon.get("clouds", {}).get(cloud_key)
    if cloud_config is None:
        raise ValueError(
            f"{addon['display_name']} is not available on {cloud_key or 'this cloud'}"
        )
    eligible_tiers = [value.upper() for value in cloud_config["eligible_tiers"]]
    if tier_key not in eligible_tiers:
        required = " or ".join(eligible_tiers)
        raise ValueError(
            f"{addon['display_name']} requires {required} tier on {cloud_key}"
        )

    as_of = _pricing_date(pricing_date)
    standard_rate = float(cloud_config["standard_rate_pct"])
    applied_rate = standard_rate
    promotion = cloud_config.get("promotion")
    active_promotion = None
    if promotion:
        promotion_end = date.fromisoformat(promotion["end_date"])
        if as_of <= promotion_end:
            applied_rate = float(promotion["rate_pct"])
            active_promotion = {
                **promotion,
                "rate_pct": applied_rate,
            }

    return {
        "addon_type": addon_key,
        "display_name": addon["display_name"],
        "sku": addon["sku"],
        "cloud": cloud_key,
        "tier": tier_key,
        "standard_rate_pct": standard_rate,
        "applied_rate_pct": applied_rate,
        "promotion": active_promotion,
        "basis": catalog["basis"],
        "as_of_date": as_of.isoformat(),
        "source_url": catalog["source_url"],
    }


def validate_platform_addon_selection(
    discount_config: dict[str, Any] | None,
    cloud: str,
    tier: str,
) -> str | None:
    """Validate a saved selection against its estimate cloud and tier."""
    selected = get_selected_platform_addon(discount_config)
    if selected:
        get_platform_addon_rate(selected, cloud, tier)
    return selected


def calculate_platform_addon_cost(
    product_spend_at_list: float,
    addon_type: str,
    cloud: str,
    tier: str,
    pricing_date: date | str | None = None,
    discount_pct: float = 0,
) -> dict[str, Any]:
    """Calculate an add-on from pre-discount Databricks product spend."""
    spend = float(product_spend_at_list)
    if not math.isfinite(spend) or spend < 0:
        raise ValueError("product_spend_at_list must be a non-negative finite number")
    discount = float(discount_pct)
    if not math.isfinite(discount) or discount < 0 or discount > 100:
        raise ValueError("discount_pct must be between 0 and 100")

    rate = get_platform_addon_rate(
        addon_type,
        cloud,
        tier,
        pricing_date=pricing_date,
    )
    standard_cost = spend * rate["standard_rate_pct"] / 100
    cost_before_discount = spend * rate["applied_rate_pct"] / 100
    discount_amount = cost_before_discount * discount / 100
    applied_cost = cost_before_discount - discount_amount
    return {
        **rate,
        "product_spend_at_list": spend,
        "standard_cost": standard_cost,
        "cost_before_discount": cost_before_discount,
        "discount_pct": discount,
        "discount_amount": discount_amount,
        "cost": applied_cost,
    }

File: frontend/src/utils/platformAddons.ts
import type {
  PlatformAddonCatalog,
  PlatformAddonDefinition,
} from './pricingBundle'
import type { PlatformAddonType } from '../types'

export type { PlatformAddonType } from '../types'

export const PLATFORM_ADDON_TYPES: PlatformAddonType[] = [
  'ENHANCED_SECURITY_COMPLIANCE',
  'MISSION_CRITICAL',
]

export interface PlatformAddonRate {
  addonType: PlatformAddonType
  displayName: string
  sku: string
  standardRatePct: number
  appliedRatePct: number
  promotionLabel?: string
  promotionEndDate?: string
  sourceUrl: string
}

export interface PlatformAddonCost extends PlatformAddonRate {
  productSpendAtList: number
  standardCost: number
  costBeforeDiscount: number
  discountPct: number
  discountAmount: number
  cost: number
}

const normalize = (value: string): string => value.toUpperCase()

export function getPlatformAddonDefinition(
  catalog: PlatformAddonCatalog,
  addonType: PlatformAddonType,
): PlatformAddonDefinition | null {
  return catalog.addons[addonType] ?? null
}

export function getPlatformAddonAvailabilityError(
  catalog: PlatformAddonCatalog,
  addonType: PlatformAddonType,
  cloud: string,
  tier: string,
): string | null {
  const definition = getPlatformAddonDefinition(catalog, addonType)
  if (!definition) return 'Platform add-on pricing is unavailable'

  const cloudKey = normalize(cloud)
  const tierKey = normalize(tier)
  const cloudConfig = definition.clouds[cloudKey]
  if (!cloudConfig) {
    return `${definition.display_name} is not available on ${cloudKey}`
  }
  if (!cloudConfig.eligible_tiers.map(normalize).includes(tierKey)) {
    return `${definition.display_name} requires ${cloudConfig.eligible_tiers.join(' or ')} tier on ${cloudKey}`
  }
  return null
}

export function getPlatformAddonRate(
  catalog: PlatformAddonCatalog,
  addonType: PlatformAddonType,
  cloud: string,
  tier: string,
  pricingDate: Date = new Date(),
): PlatformAddonRate | null {
  if (getPlatformAddonAvailabilityError(catalog, addonType, cloud, tier)) {
    return null
  }

  const definition = catalog.addons[addonType]
  const cloudConfig = definition.clouds[normalize(cloud)]
  const standardRatePct = cloudConfig.standard_rate_pct
  let appliedRatePct = standardRatePct
  let promotionLabel: string | undefined
  let promotionEndDate: string | undefined

  if (cloudConfig.promotion) {
    const endDate = new Date(`${cloudConfig.promotion.end_date}T23:59:59Z`)
    if (pricingDate.getTime() <= endDate.getTime()) {
      appliedRatePct = cloudConfig.promotion.rate_pct
      promotionLabel = cloudConfig.promotion.label
      promotionEndDate = cloudConfig.promotion.end_date
    }
  }

  return {
    addonType,
    displayName: definition.display_name,
    sku: definition.sku,
    standardRatePct,
    appliedRatePct,
    promotionLabel,
    promotionEndDate,
    sourceUrl: catalog.source_url,
  }
}

export function calculatePlatformAddonCost(
  catalog: PlatformAddonCatalog,
  addonType: PlatformAddonType | null,
  cloud: string,
  tier: string,
  productSpendAtList: number,
  discountPct: number = 0,
  pricingDate: Date = new Date(),
): PlatformAddonCost | null {
  if (!addonType) return null
  const rate = getPlatformAddonRate(
    catalog,
    addonType,
    cloud,
    tier,
    pricingDate,
  )
  if (!rate) return null

  const spend = Number.isFinite(productSpendAtList)
    ? Math.max(0, productSpendAtList)
    : 0
  const discount = Number.isFinite(discountPct)
    ? Math.min(100, Math.max(0, discountPct))
    : 0
  const costBeforeDiscount = spend * rate.appliedRatePct / 100
  const discountAmount = costBeforeDiscount * discount / 100
  return {
    ...rate,
    productSpendAtList: spend,
    standardCost: spend * rate.standardRatePct / 100,
    costBeforeDiscount,
    discountPct: discount,
    discountAmount,
    cost: costBeforeDiscount - discountAmount,
  }
}

export function getPlatformAddonDiscountPct(
  discountConfig: unknown,
): number {
  if (!discountConfig || typeof discountConfig !== 'object') return 0
  const globalDiscounts = (discountConfig as Record<string, unknown>).global
  if (!globalDiscounts || typeof globalDiscounts !== 'object') return 0
  const value = Number(
    (globalDiscounts as Record<string, unknown>).platform_addon_discount ?? 0,
  )
  return Number.isFinite(value) ? Math.min(100, Math.max(0, value)) : 0
}
